Review jdo dao

Egoes98 opened the pull request:
In this brach I have created the JDO for review and linked with the hotel table, show each hotel has many review. Also I have added the DAO methods for adding a new review and for deleting it. In case you want to retrive the review you retrive a hotel and with that hotel goes all they review it has linked. Also there is a Junit test where the DAO methods are tested.

Egoes98 committed:
The Review class

The review class and started JDO and DAO

Egoes98 committed:
Review model changes

Little changes is not finisehd yet.

Egoes98 committed:
Starting the test

The JDO works nad i have started the test but it gives null pointer exception so is not finished yet.

Egoes98 committed:
Test Finished

The test is finisehd i have to do some minor changes but is working fine.

Egoes98 committed:
Minor Changes

Deletete on cascade must be done for review to work perfectly.

Egoes98 committed:
Delete Review

Now the review deletes when you delete a hotel and also it deletes without having to delete the hotel. Test are done and wotking fine. I have done some changes in HotelDAO because it wasnt workin fine.

Egoes98 committed:
Merge branch 'master' into ReviewJDO-DAO

Egoes98 committed:
Loggers fixed

Egoes98 committed:
Some conflicting imports erased

Egoes98 committed:
Documentaion

iapellaniz reviewed (changes requested):
A couple of things:

- There's a lot of uncommented code
- Getting an error when running `ReviewDBTest`:

    ````java
    java.sql.SQLIntegrityConstraintViolationException: Duplicate entry '1' for key 'PRIMARY'
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:117)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:970)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1109)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1057)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1377)
	at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdate(ClientPreparedStatement.java:1042)
	at org.datanucleus.store.rdbms.datasource.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:98)
	at org.datanucleus.store.rdbms.datasource.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:98)
	at org.datanucleus.store.rdbms.ParamLoggingPreparedStatement.executeUpdate(ParamLoggingPreparedStatement.java:393)
	at org.datanucleus.store.rdbms.SQLController.executeStatementUpdate(SQLController.java:427)
	at org.datanucleus.store.rdbms.request.InsertRequest.execute(InsertRequest.java:414)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.insertObjectInTable(RDBMSPersistenceHandler.java:162)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.insertObject(RDBMSPersistenceHandler.java:138)
	at org.datanucleus.state.StateManagerImpl.internalMakePersistent(StateManagerImpl.java:4569)
	at org.datanucleus.state.StateManagerImpl.makePersistent(StateManagerImpl.java:4546)
	at org.datanucleus.ExecutionContextImpl.persistObjectInternal(ExecutionContextImpl.java:2026)
	at org.datanucleus.ExecutionContext.persistObjectInternal(ExecutionContext.java:339)
	at org.datanucleus.store.types.SCOUtils.validateObjectForWriting(SCOUtils.java:1220)
	at org.datanucleus.store.rdbms.scostore.ElementContainerStore.validateElementForWriting(ElementContainerStore.java:290)
	at org.datanucleus.store.rdbms.scostore.FKListStore.validateElementForWriting(FKListStore.java:958)
	at org.datanucleus.store.rdbms.scostore.FKListStore.internalAdd(FKListStore.java:562)
	at org.datanucleus.store.rdbms.scostore.AbstractListStore.add(AbstractListStore.java:113)
	at org.datanucleus.store.types.wrappers.backed.ArrayList.add(ArrayList.java:692)
	at es.deusto.spq.server.data.jdo.Hotel.addReview(Hotel.java:39)
	at es.deusto.spq.server.data.dao.HotelDAO.storeReview(HotelDAO.java:169)
	at es.deusto.spq.ReviewDBTest.reviewDAOTest(ReviewDBTest.java:43)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
    ````
    - The test should clean up whatever it changed from the database when finished
- Merge `master` so that we're sure it works with the current code base 
    That'll also allow us to run the JaCoCo report and check coverage

iapellaniz commented on src/main/java/es/deusto/spq/server/data/jdo/User.java:
```diff
@@ -18,12 +19,16 @@
 	@Persistent(defaultFetchGroup="true")
 	private String password;
 	
+	/*@Persistent(defaultFetchGroup="true", mappedBy="user", dependentElement = "true")
```
Please remove unused code

iapellaniz commented on src/main/java/es/deusto/spq/server/data/jdo/User.java:
```diff
@@ -18,12 +19,16 @@
 		this.name = name;
 		this.email = email;
 		this.password = password;
+		//reviews = new ArrayList<Review>();
```
Please remove unused code

Egoes98 replied:
> A couple of things:
> 
> * There's a lot of uncommented code
> * Getting an error when running `ReviewDBTest`:
>   ```java
>   java.sql.SQLIntegrityConstraintViolationException: Duplicate entry '1' for key 'PRIMARY'
>   at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:117)
>   at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:97)
>   at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
>   at com.mysql.cj.jdbc.ClientPreparedStatement.executeInternal(ClientPreparedStatement.java:970)
>   at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1109)
>   at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdateInternal(ClientPreparedStatement.java:1057)
>   at com.mysql.cj.jdbc.ClientPreparedStatement.executeLargeUpdate(ClientPreparedStatement.java:1377)
>   at com.mysql.cj.jdbc.ClientPreparedStatement.executeUpdate(ClientPreparedStatement.java:1042)
>   at org.datanucleus.store.rdbms.datasource.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:98)
>   at org.datanucleus.store.rdbms.datasource.dbcp2.DelegatingPreparedStatement.executeUpdate(DelegatingPreparedStatement.java:98)
>   at org.datanucleus.store.rdbms.ParamLoggingPreparedStatement.executeUpdate(ParamLoggingPreparedStatement.java:393)
>   at org.datanucleus.store.rdbms.SQLController.executeStatementUpdate(SQLController.java:427)
>   at org.datanucleus.store.rdbms.request.InsertRequest.execute(InsertRequest.java:414)
>   at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.insertObjectInTable(RDBMSPersistenceHandler.java:162)
>   at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.insertObject(RDBMSPersistenceHandler.java:138)
>   at org.datanucleus.state.StateManagerImpl.internalMakePersistent(StateManagerImpl.java:4569)
>   at org.datanucleus.state.StateManagerImpl.makePersistent(StateManagerImpl.java:4546)
>   at org.datanucleus.ExecutionContextImpl.persistObjectInternal(ExecutionContextImpl.java:2026)
>   at org.datanucleus.ExecutionContext.persistObjectInternal(ExecutionContext.java:339)
>   at org.datanucleus.store.types.SCOUtils.validateObjectForWriting(SCOUtils.java:1220)
>   at org.datanucleus.store.rdbms.scostore.ElementContainerStore.validateElementForWriting(ElementContainerStore.java:290)
>   at org.datanucleus.store.rdbms.scostore.FKListStore.validateElementForWriting(FKListStore.java:958)
>   at org.datanucleus.store.rdbms.scostore.FKListStore.internalAdd(FKListStore.java:562)
>   at org.datanucleus.store.rdbms.scostore.AbstractListStore.add(AbstractListStore.java:113)
>   at org.datanucleus.store.types.wrappers.backed.ArrayList.add(ArrayList.java:692)
>   at es.deusto.spq.server.data.jdo.Hotel.addReview(Hotel.java:39)
>   at es.deusto.spq.server.data.dao.HotelDAO.storeReview(HotelDAO.java:169)
>   at es.deusto.spq.ReviewDBTest.reviewDAOTest(ReviewDBTest.java:43)
>   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>   at java.base/java.lang.reflect.Method.invoke(Method.java:564)
>   at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
>   at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
>   at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
>   at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
>   at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
>   at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
>   at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
>   at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
>   at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
>   at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
>   at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
>   at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
>   at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
>   at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
>   at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
>   at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
>   at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
>   at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
>   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>   at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>   at java.base/java.lang.reflect.Method.invoke(Method.java:564)
>   at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
>   at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
>   at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
>   at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
>   at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
>   ```
>   
>   
>   
>   * The test should clean up whatever it changed from the database when finished
> * Merge `master` so that we're sure it works with the current code base
>   That'll also allow us to run the JaCoCo report and check coverage

In my case the test works fine. Master is alredy merged and the test erases the data it uses.

iapellaniz replied:
Yes, don't worry about the Exception, it was a problem on my end!

Egoes98 committed:
Documentation and removed unused code

I removed the used code, thank Iñigo i didnt remember i had that code. Aslo i have added some documentation if you think there should be more please tell my what methods do you think are without documentation. I have to solve my DB problems and them i will see what error it gives. The other day worked fine for me.

iker-barriocanal reviewed (dismissed):
## Some notes:
* `Review`
  * It isn't documented
  * What's the maximum score? and minimum?
  * Can a user write as many reviews as he wants?
* `ReviewDBTest`
  * It is not in the correct path.
  * To compare if two reviews are the same, instead of checking every single attribute it has, check out `equals` method.
- `HotelDAO`
  - Review-related stuff is placed in the `HotelDAO`. Why doesn't a `ReviewDAO` exists?
  - What is the ID of a review for? There's no way to retrieve a review with the ID.

Egoes98 committed:
Changed how reviews work

I have changed nearly all the way reviews work. I have done it with foreign keys and now hotels and users dont store reviews you ahve to get them from DB. In the test i have tested the store reviews, get reviews that a user posted, get reviews about a hotel and delete hotel. it works fine on my pc. I should add some more documentation, i would add it tomorrow. As you requested I have done a ReviewDAO. Now a user can only make one review per hotel.  I have  moved the test file. Refering the ID i use it for deleting the review and in the future if you see it necessary(i dont thinks it is) we coud retrieve the review by id.

iker-barriocanal commented on src/main/java/es/deusto/spq/server/data/jdo/Review.java:
```diff
@@ -0,0 +1,95 @@
+
+	@Override
+	public boolean equals(Object obj) {
+		//Two Review are equal objects if and only if they have the same userID.
```
Are actually two reviews the same if they have been made by the same user?

Egoes98 replied:
This is a mistake i have done i have to correct it. The user is the same if the reviewID is the same.

iapellaniz replied:
Seems fine now

Egoes98 committed:
Some changes and Documentations.

I have added more documentation still i see in the reports i should add more.

Egoes98 committed:
Documentation mistake

A documentation mistake i have done.

Egoes98 committed:
Clean the code

iapellaniz reviewed (changes requested):
Sorry, I didn't notice yesterday, but the `ReviewDTO` is missing.

Again, hate to be this picky, but the `ReviewDAO` class (one of the most important classes to comment) is not documented.

Other than that, looks pretty good.

Egoes98 committed:
Added The DTO

I have added the DTO. I will tomorrow look closely at the reports and see what parts of the reviewDTO need to be comented.

Egoes98 committed:
More Documentation(Still i think some documentation needed)

I ahve added documentation but still i should add something more.

Egoes98 committed:
Docuementation

I think documentation is done. I have checked the Checklist results and looks like Javadock is corrected.

Egoes98 committed:
Merge branch 'master' into ReviewJDO-DAO

Egoes98 committed:
Revert "Merge branch 'master' into ReviewJDO-DAO"

This reverts commit 863353ad82cf9835341940848836b2c7cdd1be22, reversing
changes made to fe770688f717674cf2388eb42fed5e591303f28f.

Egoes98 committed:
Travis JDK 10

Egoes98 committed:
Revert "Revert "Merge branch 'master' into ReviewJDO-DAO""

This reverts commit a7088b747909fefd8ac047050bb5729590c0aca9.

Egoes98 commented:
Sorry i have had problems when merging because i had eclipse configured for another project and it changed a parameter to final so i had problems.

Egoes98 committed:
Sorry, all workin fine. conflict solved.

I am relly sorry for this last commits but sclipse was messing this up and i had a bad time solving this stupid thing.

Egoes98 committed:
Rmoved the changes eclipse did.

Some finals and the public and private statements in the Interface sorry about all of this.
